pennylane/transforms/batch_input.py

```diff
@@ -73,25 +72,39 @@ def circuit(inputs, weights):
     array([0.46230079, 0.73971315, 0.95666004, 0.5355225 , 0.66180948,
             0.44519553, 0.93874261, 0.9483197 , 0.78737918, 0.90866411])>
     """
-    parameters = tape.get_parameters(trainable_only=False)
-
     argnum = tuple(argnum) if isinstance(argnum, (list, tuple)) else (int(argnum),)
 
-    non_trainable, trainable = [], []
-    for idx, param in enumerate(parameters):
-        if idx in argnum:
-            non_trainable.append(param)
-        else:
-            trainable.append(param)
+    trainable_params = tape.get_parameters()
+    all_parameters = sum(
+        (op.parameters if op.parameters != [] else [None] for op in tape.operations), []
+    )
+    argnum_params = [all_parameters[i] for i in argnum]
 
-    if len(np.unique([qml.math.shape(x)[0] for x in non_trainable])) != 1:
+    if any(
+        any(param is train_param for train_param in trainable_params) for param in argnum_params
+    ):
+        raise ValueError(
+            "Batched inputs must be non-trainable. Please make sure that the parameters indexed by "
+            + "'argnum' have 'requires_grad' set to False."
+        )
+
+    if len(np.unique([qml.math.shape(x)[0] for x in argnum_params])) != 1:
         raise ValueError(
             "Batch dimension for all gate arguments specified by 'argnum' must be the same."
         )
 
+    batch_size = qml.math.shape(argnum_params[0])[0]
+
     outputs = []
-    for inputs in zip(*non_trainable):
-        outputs += [list(inputs) + trainable]
+    for i in range(batch_size):
+        batch = []
+        for idx, param in enumerate(all_parameters):
+            if param is None:
+                continue
+            if idx in argnum:
+                param = param[i]
```

Comment:
`idx` and `argnum` are iterating over parameters, right? So if there is no parameter, I suppose the counter should not go up. Accordingly, the approach using `None` as dummy parameter list for parameter-free operations could be skipped :)
Indeed, the following example breaks with the current PR version (it's the docstring with one additional `Hadamard` gate)
```py
dev = qml.device("default.qubit", wires = 2, shots=None)

@qml.batch_input(argnum=1)
@qml.qnode(dev, diff_method="parameter-shift", interface="tf")
def circuit(inputs, weights):
    qml.RY(weights[0], wires=0)
    qml.Hadamard(wires=0)
    qml.AngleEmbedding(inputs, wires = range(2), rotation="Y")
    qml.RY(weights[1], wires=1)
    return qml.expval(qml.PauliZ(1))

x = tf.random.uniform((10, 2), 0, 1)
w = tf.random.uniform((2,), 0, 1)
circuit(x, w)
```

Reply:
Right now `argnum` is defined as the index indicating the location of the operation that uses the batched inputs. In this example by using `argnum=1` you are saying that the `Hadamard` will use the batched inputs. The code crashes because it tries to find the shape of `None`.

Reply:
Ah I see! I think this is a bit different than the usage of `argnum` across the Pennylane codebase, as far as I can tell. That is, parameter-free operations typically are not counted when using/interpreting `argnum`.
Unfortunately, there are a few inconsistencies, in particular with transforms: When applying them to QNodes, they sometimes refer to QNode arguments, and sometimes to Tape arguments. However, as far as I am aware, they do not refer to tape operations. Maybe would be good to confirm with someone, still :)

Reply:
Thanks for explaining this! I didn't know it. I added the changes that you proposed such that only parametric ops are taken into account when defining `argnum`.

Reply:
Brilliant, thank you a lot! 
As I mentioned, this could be a bit more consistent and well-documented across the codebase, so sorry for off-loading the adaptation to the current state :grimacing: